### ISSUE TITLE
Incorrect (or absent) error message for DefaultDict

### ISSUE THREAD

**Issue by ilevkivskyi** [python/mypy#2986]
With this code:
```python
from typing import DefaultDict
from collections import defaultdict

d: DefaultDict[str, int]
d[1] = 2
```
mypy shows a strange error (note ``"dict"`` instead of ``"DefaultDict"``):
```
error: Invalid index type "int" for "dict"; expected type "str"
```
Moreover, if I remove the (redundant) import from ``collections``, then mypy passes without errors.

**Comment by ilevkivskyi** [comment]
Since it looks like this bug was introduced by my PR, I will fix this. You could assign this issue to me (I can't do this myself).

**Comment by ilevkivskyi** [comment]
This was partially fixed by #3147, now the error message correctly mentions ``defaultdict[str, int]`` instead of just ``dict``. However, still the message disappears if redundant ``collections`` import is removed.

**Comment by ilevkivskyi** [comment]
This is now fixed by https://github.com/python/typeshed/issues/1456
